airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/common.py

```diff
@@ -0,0 +1,255 @@
+    │       ├── Profiles
+    │       └── SubProfilesStream
+    │           └── PaginationStream
+    │               ├── SponsoredDisplayAdGroups
+    │               ├── SponsoredDisplayCampaigns
+    │               ├── SponsoredDisplayProductAds
+    │               ├── SponsoredDisplayTargetings
+    │               ├── SponsoredProductAdGroups
+    │               ├── SponsoredProductAds
+    │               ├── SponsoredProductCampaigns
+    │               ├── SponsoredProductKeywords
+    │               ├── SponsoredProductNegativeKeywords
+    │               ├── SponsoredProductTargetings
+    │               ├── SponsoredBrandsCampaigns
+    │               ├── SponsoredBrandsAdGroups
+    │               ├── SponsoredBrandsKeywords
+    │               ├── SponsoredDisplayReportStream
+    │               ├── SponsoredProductsReportStream
+    │               └── SponsoredBrandsReportStream
+    └── ReportStream
+        ├── SponsoredBrandsReportStream
+        ├── SponsoredDisplayReportStream
+        └── SponsoredProductsReportStream
+
+BasicAmazonAdsStream is base class inherited from CDK's Stream class and used
+for storing list of profiles that later be used by all the streams to get
+profile id. Also it stores pydantic model and API url for requests.
+
+AmazonAdsStream is Http based class, it used for making request that could be
+accomlished by single http call (any but report streams). Also in contains
+transform method to modify output data (see transform method comment).
+
+SubProfilesStream is subclass for http streams to perform read_records from
+basic class for EACH profile from self._profiles list.
+
+PaginationStream class implement Amazon Ads API pagintaion for all requests.
+This is base class for all the sync http streams that used by source.
+
+ReportStream (It implemented on report_stream.py file) is subclass for async
+report streams. This is not standard http stream and used for generating
+reports for profiles from BasicAmazonAdsStream _profiles list.
+
+"""
+
+
+class ErrorResponse(BaseModel):
+    code: str
+    details: str
+    requestId: str
+
+
+class BasicAmazonAdsStream(Stream, ABC):
+    """
+    Base class for all Amazon Ads streams.
+    """
+
+    def __init__(self, config, profiles: List[Profile] = None):
+        self._profiles = profiles or []
+        self._client_id = config.client_id
+        self._url = config.host or URL_BASE
+
+    @property
+    @abstractmethod
+    def model(self) -> CatalogModel:
+        """
+        Pydantic model to represent json schema
+        """
+
+    def get_json_schema(self):
+        return self.model.schema()
+
+
+# Basic full refresh stream
+class AmazonAdsStream(HttpStream, BasicAmazonAdsStream):
+    """
+    Class for getting data from streams that based on single http request.
+    """
+
+    # List of properties that used by transform method to convert "object"
+    # subtype with complex dict to plan string.
+    flatten_properties = []
```

Comment:
instead of flattening these properties, can we instead drop them from the record? 

The reason is that if we provide them as a string today, then in the future if we want to fully support them, users will need to run a schema migration or potentially re-import data to change the data types in the destination. Instead, let's just not provide it at all and wait for a user to request it then support it "properly"

Reply:
Done